File: data/data-proc/src/data_proc/source_reader.py
from __future__ import annotations

import json
import zipfile
from pathlib import Path

import click

from data_proc.utils import bible_sources
from data_proc.utils.bible_tandem import TandemBible

REPO_ROOT = Path(__file__).resolve().parents[4]
DEFAULT_ENGLISH_XML = REPO_ROOT / bible_sources.DEFAULT_ENGLISH_COLLECTION
DEFAULT_HEBREW_ZIP = REPO_ROOT / bible_sources.DEFAULT_HEBREW_ZIP
DEFAULT_OUT_FILE = REPO_ROOT / "source.zip"

BOOK_SLUGS = {
    "GEN": "genesis",
    "EXO": "exodus",
    "LEV": "leviticus",
    "NUM": "numbers",
    "DEU": "deuteronomy",
    "JOS": "joshua",
    "JDG": "judges",
    "RUT": "ruth",
    "1SA": "1-samuel",
    "2SA": "2-samuel",
    "1KI": "1-kings",
    "2KI": "2-kings",
    "1CH": "1-chronicles",
    "2CH": "2-chronicles",
    "EZR": "ezra",
    "NEH": "nehemiah",
    "EST": "esther",
    "JOB": "job",
    "PSA": "psalms",
    "PRO": "proverbs",
    "ECC": "ecclesiastes",
    "SON": "song-of-songs",
    "ISA": "isaiah",
    "JER": "jeremiah",
    "LAM": "lamentations",
    "EZE": "ezekiel",
    "DAN": "daniel",
    "HOS": "hosea",
    "JOE": "joel",
    "AMO": "amos",
    "OBA": "obadiah",
    "JON": "jonah",
    "MIC": "micah",
    "NAH": "nahum",
    "HAB": "habakkuk",
    "ZEP": "zephaniah",
    "HAG": "haggai",
    "ZEC": "zechariah",
    "MAL": "malachi",
}


def _chapter_archive_name(slug: str, chapter: int) -> str:
    return f"{slug}/chapter{chapter}.json"


def _directory_zip_info(path: str) -> zipfile.ZipInfo:
    info = zipfile.ZipInfo(path)
    info.external_attr = 0o755 << 16
    info.compress_type = zipfile.ZIP_STORED
    return info


def _build_index_payload(
    tandem: TandemBible,
    *,
    included_book_codes: set[str],
) -> dict:
    books = []

    for book_code, book_en, book_he in tandem.iter_books():
        if book_code not in included_book_codes:
            continue

        chapter_count = max(
            (chapter for _, chapter in tandem.iter_chapters(book_filter=book_code)),
            default=0,
        )
        books.append(
            {
                "code": book_code,
                "slug": BOOK_SLUGS[book_code],
                "en": book_en,
                "he": book_he,
                "chapter_count": chapter_count,
            }
        )

    books.sort(key=lambda item: bible_sources.BOOK_ORDER.get(item["code"], 999))
    return {"books": books}


def build_source_reader_archive(
    tandem: TandemBible,
    out_file: Path,
    *,
    book_filter: str | None = None,
    chapter_filter: int | None = None,
) -> tuple[int, dict]:
    chapters_written = 0
    book_key = (book_filter or "").strip()
    included_book_codes: set[str] = set()

    out_file.parent.mkdir(parents=True, exist_ok=True)

    with zipfile.ZipFile(out_file, "w", compression=zipfile.ZIP_DEFLATED) as archive:
        for book_code, book_en, book_he in tandem.iter_books(book_filter=book_key):
            slug = BOOK_SLUGS[book_code]
            wrote_book_directory = False

            for _, chapter in tandem.iter_chapters(book_filter=book_code):
                if chapter_filter is not None and chapter != chapter_filter:
                    continue

                verses = [
                    {
                        "verse": verse.verse,
                        "en": verse.en_raw,
                        "he": verse.he_clean,
                    }
                    for verse in tandem.iter_verses(book_code, chapter)
                ]
                if not verses:
                    continue

                payload = {
                    "book_code": book_code,
                    "slug": slug,
                    "book": book_en,
                    "book_he": book_he,
                    "chapter": chapter,
                    "verses": verses,
                }
                if not wrote_book_directory:
                    archive.writestr(_directory_zip_info(f"{slug}/"), "")
                    wrote_book_directory = True
                archive.writestr(
                    _chapter_archive_name(slug, chapter),
                    json.dumps(payload, ensure_ascii=False, separators=(",", ":")),
                )
                included_book_codes.add(book_code)
                chapters_written += 1

        index_payload = _build_index_payload(tandem, included_book_codes=included_book_codes)
        archive.writestr("index.json", json.dumps(index_payload, ensure_ascii=False, separators=(",", ":")))

    return chapters_written, index_payload


@click.command("build-source-reader")
@click.option("--out-file", type=click.Path(path_type=Path, dir_okay=False), default=DEFAULT_OUT_FILE, show_default=True)
@click.option("--english-xml", type=click.Path(path_type=Path, exists=True, dir_okay=False), default=DEFAULT_ENGLISH_XML, show_default=True)
@click.option("--hebrew-zip", type=click.Path(path_type=Path, exists=True, dir_okay=False), default=DEFAULT_HEBREW_ZIP, show_default=True)
@click.option("--book", "book_filter", default=None)
@click.option("--chapter", "chapter_filter", type=int, default=None)
def build_source_reader_command(
    out_file: Path,
    english_xml: Path,
    hebrew_zip: Path,
    book_filter: str | None,
    chapter_filter: int | None,
) -> None:
    tandem = TandemBible.load(english_xml=english_xml, hebrew_zip=hebrew_zip)
    chapters_written, index_payload = build_source_reader_archive(
        tandem,
        out_file,
        book_filter=book_filter,
        chapter_filter=chapter_filter,
    )
    click.echo(f"Wrote {chapters_written} source-reader chapter files across {len(index_payload['books'])} books to {out_file}.")

File: src/lib/sourceReader.ts
import type { Lang, SourceRef } from "../types";
import { getSearchWithLanguage } from "./language";

export type SourceBook = {
  code: string;
  slug: string;
  en: string;
  he: string;
};

export type SourceIndexBook = {
  code: string;
  slug: string;
  en: string;
  he: string;
  chapter_count: number;
};

export type SourceIndexPayload = {
  books: SourceIndexBook[];
};

export type SourceVerse = {
  verse: number;
  en: string;
  he: string;
};

export type SourceChapterPayload = {
  book_code: string;
  slug: string;
  book: string;
  book_he: string;
  chapter: number;
  verses: SourceVerse[];
};

const SOURCE_BOOK_DEFS = [
  { code: "GEN", slug: "genesis", en: "Genesis", he: "בראשית" },
  { code: "EXO", slug: "exodus", en: "Exodus", he: "שמות" },
  { code: "LEV", slug: "leviticus", en: "Leviticus", he: "ויקרא" },
  { code: "NUM", slug: "numbers", en: "Numbers", he: "במדבר" },
  { code: "DEU", slug: "deuteronomy", en: "Deuteronomy", he: "דברים" },
  { code: "JOS", slug: "joshua", en: "Joshua", he: "יהושע" },
  { code: "JDG", slug: "judges", en: "Judges", he: "שופטים" },
  { code: "RUT", slug: "ruth", en: "Ruth", he: "רות" },
  { code: "1SA", slug: "1-samuel", en: "1 Samuel", he: "שמואל א" },
  { code: "2SA", slug: "2-samuel", en: "2 Samuel", he: "שמואל ב" },
  { code: "1KI", slug: "1-kings", en: "1 Kings", he: "מלכים א" },
  { code: "2KI", slug: "2-kings", en: "2 Kings", he: "מלכים ב" },
  { code: "1CH", slug: "1-chronicles", en: "1 Chronicles", he: "דברי הימים א" },
  { code: "2CH", slug: "2-chronicles", en: "2 Chronicles", he: "דברי הימים ב" },
  { code: "EZR", slug: "ezra", en: "Ezra", he: "עזרא" },
  { code: "NEH", slug: "nehemiah", en: "Nehemiah", he: "נחמיה" },
  { code: "EST", slug: "esther", en: "Esther", he: "אסתר" },
  { code: "JOB", slug: "job", en: "Job", he: "איוב" },
  { code: "PSA", slug: "psalms", en: "Psalms", he: "תהילים" },
  { code: "PRO", slug: "proverbs", en: "Proverbs", he: "משלי" },
  { code: "ECC", slug: "ecclesiastes", en: "Ecclesiastes", he: "קהלת" },
  { code: "SON", slug: "song-of-songs", en: "Song of Songs", he: "שיר השירים" },
  { code: "ISA", slug: "isaiah", en: "Isaiah", he: "ישעיהו" },
  { code: "JER", slug: "jeremiah", en: "Jeremiah", he: "ירמיהו" },
  { code: "LAM", slug: "lamentations", en: "Lamentations", he: "איכה" },
  { code: "EZE", slug: "ezekiel", en: "Ezekiel", he: "יחזקאל" },
  { code: "DAN", slug: "daniel", en: "Daniel", he: "דניאל" },
  { code: "HOS", slug: "hosea", en: "Hosea", he: "הושע" },
  { code: "JOE", slug: "joel", en: "Joel", he: "יואל" },
  { code: "AMO", slug: "amos", en: "Amos", he: "עמוס" },
  { code: "OBA", slug: "obadiah", en: "Obadiah", he: "עובדיה" },
  { code: "JON", slug: "jonah", en: "Jonah", he: "יונה" },
  { code: "MIC", slug: "micah", en: "Micah", he: "מיכה" },
  { code: "NAH", slug: "nahum", en: "Nahum", he: "נחום" },
  { code: "HAB", slug: "habakkuk", en: "Habakkuk", he: "חבקוק" },
  { code: "ZEP", slug: "zephaniah", en: "Zephaniah", he: "צפניה" },
  { code: "HAG", slug: "haggai", en: "Haggai", he: "חגי" },
  { code: "ZEC", slug: "zechariah", en: "Zechariah", he: "זכריה" },
  { code: "MAL", slug: "malachi", en: "Malachi", he: "מלאכי" },
] as const;

export const SOURCE_BOOKS: readonly SourceBook[] = SOURCE_BOOK_DEFS;

const BOOK_BY_CODE = new Map(SOURCE_BOOKS.map((book) => [book.code, book]));
const ORDER_BY_SLUG = new Map(SOURCE_BOOKS.map((book, index) => [book.slug, index]));

export type ReaderNeighbor = {
  bookSlug: string;
  chapter: number;
};

export function getSourceBookByCode(bookCode: string | undefined): SourceBook | undefined {
  return bookCode ? BOOK_BY_CODE.get(bookCode) : undefined;
}

export function buildVerseHash(verse: number): string {
  return `#v${verse}`;
}

export function buildReaderPath(bookSlug?: string, chapter?: number): string {
  if (!bookSlug) return "/read";
  if (chapter === undefined) return `/read/${bookSlug}`;
  return `/read/${bookSlug}/${chapter}`;
}

export function buildReaderHref(params: {
  lang: Lang;
  bookSlug: string;
  chapter?: number;
  verse?: number;
}): string {
  const path = buildReaderPath(params.bookSlug, params.chapter);
  const search = getSearchWithLanguage("", params.lang);
  const hash = params.verse === undefined ? "" : buildVerseHash(params.verse);
  return `${path}${search}${hash}`;
}

export function buildReaderHrefFromSource(source: SourceRef | undefined, lang: Lang): string | null {
  if (!source?.book_code || source.chapter === undefined || source.quote_verse_start === undefined) return null;
  const book = getSourceBookByCode(source.book_code);
  if (!book) return null;
  return buildReaderHref({ lang, bookSlug: book.slug, chapter: source.chapter, verse: source.quote_verse_start });
}

export function getOrderedSourceIndexBooks(books: readonly SourceIndexBook[]): SourceIndexBook[] {
  return [...books].sort((left, right) => {
    const leftOrder = ORDER_BY_SLUG.get(left.slug) ?? Number.MAX_SAFE_INTEGER;
    const rightOrder = ORDER_BY_SLUG.get(right.slug) ?? Number.MAX_SAFE_INTEGER;
    return leftOrder - rightOrder;
  });
}

export function getReaderNeighbors(
  books: readonly SourceIndexBook[],
  bookSlug: string,
  chapter: number
): { previous: ReaderNeighbor | null; next: ReaderNeighbor | null } {
  const orderedBooks = getOrderedSourceIndexBooks(books);
  const bookIndex = orderedBooks.findIndex((book) => book.slug === bookSlug);
  if (bookIndex < 0) return { previous: null, next: null };

  const currentBook = orderedBooks[bookIndex];
  const previous =
    chapter > 1
      ? { bookSlug, chapter: chapter - 1 }
      : bookIndex > 0
        ? { bookSlug: orderedBooks[bookIndex - 1]!.slug, chapter: orderedBooks[bookIndex - 1]!.chapter_count }
        : null;
  const next =
    chapter < currentBook.chapter_count
      ? { bookSlug, chapter: chapter + 1 }
      : bookIndex < orderedBooks.length - 1
        ? { bookSlug: orderedBooks[bookIndex + 1]!.slug, chapter: 1 }
        : null;

  return { previous, next };
}
